rust_src/src/marker.rs

```diff
@@ -462,4 +491,316 @@ impl LispBufferRef {
+
+    let mut best_above = buffer_ref.z();
+    let mut best_above_byte = buffer_ref.z_byte();
+
+    // If this buffer has as many characters as bytes,
+    // each character must be one byte.
+    // This takes care of the case where enable-multibyte-characters is nil.
+    if best_above == best_above_byte {
+        return bytepos;
+    }
+
+    let mut best_below = buffer_ref.beg();
+    let mut best_below_byte = buffer_ref.beg_byte();
+
+    macro_rules! consider {
+        ($bpos:expr, $cpos:expr) => {
+            let mut changed = false;
+            if $bpos == bytepos {
+                return $cpos;
+            } else if $bpos > bytepos {
+                if $bpos < best_above_byte {
+                    best_above = $cpos;
+                    best_above_byte = $bpos;
+                    changed = true;
+                }
+            } else if $bpos > best_below_byte {
+                best_below = $cpos;
+                best_below_byte = $bpos;
+                changed = true;
+            }
+            if changed {
+                if best_above - best_below == best_above_byte - best_below_byte {
+                    return best_below + (bytepos - best_below_byte);
+                }
+            }
+        };
+    }
+
+    consider!(buffer_ref.pt_byte(), buffer_ref.pt());
+    consider!(buffer_ref.gpt_byte(), buffer_ref.gpt());
+    consider!(buffer_ref.begv_byte(), buffer_ref.begv());
+    consider!(buffer_ref.zv_byte(), buffer_ref.zv());
+
+    unsafe {
+        if b == cached_buffer && buffer_ref.modifications() == cached_modiff {
+            consider!(cached_bytepos, cached_charpos);
+        }
+    }
+
+    for m in buffer_ref.markers().iter() {
+        consider!(m.bytepos_or_error(), m.charpos_or_error());
+        // If we are down to a range of 50 chars,
+        // don't bother checking any other markers;
+        // scan the intervening chars directly now.
+        if best_above - best_below < 50 {
+            break;
+        }
+    }
+
+    // We get here if we did not exactly hit one of the known places.
+    // We have one known above and one known below.
+    // Scan, counting characters, from whichever one is closer.
+
+    if bytepos - best_below_byte < best_above_byte - bytepos {
+        let record = bytepos - best_below_byte > 5000;
+
+        while best_below_byte < bytepos {
+            best_below += 1;
+            best_below_byte = buffer_ref.inc_pos(best_below_byte);
+        }
+
+        // If this position is quite far from the nearest known position,
+        // cache the correspondence by creating a marker here.
+        // It will last until the next GC.
+        // But don't do it if BUF_MARKERS is nil;
+        // that is a signal from Fset_buffer_multibyte.
+        if record && buffer_ref.markers().is_some() {
+            unsafe {
+                build_marker(b, best_below, best_below_byte);
+            }
+        }
+        // byte_char_debug_check(b, best_below, best_below_byte);
```

Comment:
Rather than commenting out these calls, you could change `byte_char_debug_check` to return a bool and then do `debug_assert!(byte_char_debug_check(b, best_below, best_below_byte`. What do you think?